### ISSUE TITLE
Record on single line setting

### ISSUE THREAD

**Issue by nojaf** [fsprojects/fantomas#697]
## Proposal

Fantomas currently formats a record on multiple lines if there are multiple fields.
Even though the record itself can still be short.

For [example](https://jindraivanek.gitlab.io/fantomas-ui/#?code=C4TwDgpgBAglC8UDesBcUCWA7YBuKAQutsAL4CwAUFQDYTBQCGCyUAwiwIz4EsBMFapToMARixTMA7hmAALdvx4sAzFFJA)

```fsharp
type A = { A: int; B: int }

let a = { C = 1; B = 2 }

let b = { a with C = 2; B = 3 }
```

will be formatted as

```fsharp
type A =
    { A: int
      B: int }

let a =
    { C = 1
      B = 2 }

let b =
    { a with
          C = 2
          B = 3 }
```

If two or more fields are involved, every expression is put on multiple lines.
The suggestion from G-Research is to introduce a new setting to control with maximum allowed record width. And only format on multiple lines if the record exceeds this setting.

**Proposed setting name** : `MaxSingleLineRecordWidth`
**Default value** : 40

## Impact

This would introduce a breaking change stylistically speaking. 
It would require a change in [multiline](https://github.com/fsprojects/fantomas/blob/3d2e39eb838a83cc9ce3127d48e2cd65405ab656/src/Fantomas/SourceTransformer.fs#L17) and has some impact on CodePrinter.

## Gains

The outcome would be more concise code.

**Comment by jindraivanek** [comment]
+1

Actually, I would like global setting like this that would do this for all potentially multiline expressions. (but that is potentially hard to implement)

**Comment by auduchinok** [comment]
Could it maybe check if the line fits line length with single-line record formatting instead?

**Comment by nojaf** [comment]
Not sure you would always want this.
For example:

```fsharp
let a = { LongIdentifier = "foo"; X = a; BaaaarrIdentifier = "meeeeeeh"; Meh = FooAgain; A = y; Y = e; C = c; D = dd }
```

This is 118 long and would in theory fit on a PageWidth 120 setting.
However, for me, this should be on multiple lines.

I think a similar setting could also be useful for deciding when an array should be on multiple lines or not. Debatable whether this should be a separate setting or the same setting.

f.ex.
```fsharp
let a = [ "Fooooooooooooo";  "Baaaaaaaaaaaaaaaaaar"; "Meeeeeeeeeeeeeeeh"; "Turn"; "To"; "Stooooooooooooooooooooonees" ]
```

This again is 119 characters long and I would use multiple lines instead of placing it on one as it does [now](https://jindraivanek.gitlab.io/fantomas-ui/#?code=DYUwLgBAhhC8EG0ICIBiB7TXteQbghQCEpSzyKyAnfFAWREaeZYAtbkAVAVyoDsOndBwDKYHBMlY+jAM7IIAXSA).

We could introduce some new construct that tries a naive solution to format an expression and then check if that exceeds a certain length.

Something along the lines of
```fsharp

**Comment by nojaf** [comment]
Implemented in #721
